### PR TITLE
Implement Tictactoe

### PR DESCRIPTION
**Pull request by andremleblanc**
**Feature**

Add an unbeatable Tic Tac Toe implementation leveraging the existing framework.

**Description**

This is a partially complete implementation. The UI of the app can be accessed on the root of the frontend `localhost:3000`, the game can be played, and the computer can't be beaten (and may heckle you a little). I've left comments in parts of the code where I would love some feedback and I've left a list of remaining work to make this PR feature complete. 

**Testing Instructions**

1. Start the app [following the instructions in the readme](https://github.com/andremleblanc/tictactoe#getting-started).
2. Go to [http://localhost:3000/](http://localhost:3000/).
3. Play the game and ensure you can't beat the computer.

**Feature List**

- [x] Tic Tac Toe UI: Allow players to play tic tac toe through the frontend.
- [x] Unbeatable AI: Have players compete against an AI that cannot be beaten.
- [ ] Difficulty: Allow players to chose a difficulty for the AI.
- [ ] First Player Settings: Allow players to select who goes first (themselves or the AI).
- [ ] Mark the Winning Line: Show which row was the winning row.
- [ ] Tests: Ensure there is adequate test coverage.
- [ ] Edge Cases: Ensure edge cases are covered (I've covered some, like disabling squares that have been clicked, but I'd imagine there are more to cover).

### CODE REVIEW

File: packages/client/src/components/App.tsx [hunk 1]
```diff
@@ -1,7 +1,73 @@
...
 
-export const App = () => (
-  <main>
-    <div>FIXME: add your solution here!</div>
-  </main>
-);
+import { begin, move } from '../api'
+import { Square } from './Square'
+
+export class App extends React.Component<unknown, GameStatus> {
```
Comment (andremleblanc) [inline packages/client/src/components/App.tsx:9]: This is something I wanted to spend a little more time researching. What should the properties type be of the highest level Component which does not have properties.

File: packages/client/src/components/App.tsx [hunk 2]
```diff
@@ -1,7 +1,73 @@
...
+      winner: null
+    };
+  }
+
+  componentDidMount() {
+    begin(Difficulty.hard, Player.human).then(res => this.setState(res))
+  }
+
+  handleClickOnSquare(index: number) {
+    const { board } = this.state
```
Comment (andremleblanc) [inline packages/client/src/components/App.tsx:23]: I noticed some duplication with setting board in each function. It seemed like maybe that was intentional with the linting settings, but I'd be curious for feedback here.

File: packages/client/src/components/App.tsx [hunk 3]
```diff
@@ -1,7 +1,73 @@
...
-    <div>FIXME: add your solution here!</div>
-  </main>
-);
+import { begin, move } from '../api'
+import { Square } from './Square'
+
+export class App extends React.Component<unknown, GameStatus> {
+  constructor(props: unknown) {
+    super(props)
+    this.state = {
```
Comment (andremleblanc) [inline packages/client/src/components/App.tsx:12]: I'm curious if there is an idiomatic way to use the type/interface definition here to initialize state.

File: packages/client/src/components/App.tsx [hunk 4]
```diff
@@ -1,7 +1,73 @@
...
+
+  renderSquare(index: number) {
+    const { board, winner } = this.state
+    const shouldDisableButton = winner !== null || board[index] !== null
+    return (<Square isDisabled={shouldDisableButton} handleClick={() => this.handleClickOnSquare(index)} value={board[index]} />)
+  }
+
+  render() {
+    const { winner } = this.state;
+    const messageMap = {
```
Comment (andremleblanc) [inline packages/client/src/components/App.tsx:38]: I don't like that these keys are "magic". I tried a couple of approaches to reuse the `Marker` and `Winner` classes to define the keys here, but was having difficulty. Also, if we were to continue this feature to include the first player setting, then we'd want a different solution here since the roles would change.

File: packages/server/src/process_move_service.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        const simulatedBoardState = board.slice();
+        simulatedBoardState[move] = Marker.o
+        return ProcessMoveService.calculateMinimaxScore(simulatedBoardState, 0, false)
+    }
+
+    static checkForDraw(board: Board) {
+        return board.every(square => square !== null)
+    }
+
+    static checkForWinner(board: Board) {
```
Comment (andremleblanc) [inline packages/server/src/process_move_service.ts:53]: I took the intuitive approach here and didn't put much effort into optimizing performance. Since it's such a small array, I think the approach is adequate, but if it were a larger grid or we had an emphasis on performance, we'd want to look at this.

File: packages/server/src/process_move_service.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+import { WinningLines } from "./constants";
+
+export class ProcessMoveService {
+    board: Board;
+
+    constructor(board: Board) {
+        this.board = board;
+    }
+
+    static calculateMinimaxScore(board: Board, depth: number, aiTurn: boolean) {
```
Comment (andremleblanc) [inline packages/server/src/process_move_service.ts:11]: There are a few static functions in here and I'm not sure how idiomatic that is, definitely a spot I'd love some feedback. I went this direction because `calculateMinimaxScore` needed its own copy of board (not the original passed into the class) and the compiler didn't like that the function didn't reference `this`. I would think that extracting these static methods into a new class that could use instance functions would be a good step.

File: packages/server/src/process_move_service.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        const scoreMap = {
```
Comment (andremleblanc) [inline packages/server/src/process_move_service.ts:12]: I don't like the magic keys and magic values here. Again, I tried playing with this a little to reuse the existing types/interfaces but was having difficulty finding the right abstraction. Pragmatically, it might make sense to wait to work on these until more of this feature was developed.

File: packages/server/src/process_move_service.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+            O: 10,
+            X: -10,
+            tie: 0
+        }
+
+        let finalScore!: number;
```
Comment (andremleblanc) [inline packages/server/src/process_move_service.ts:18]: I added the non-null assertion here because a score is always returned from this recursive function call, but the compiler was having difficulty believing me. I'd love feedback on this approach.

File: packages/server/src/process_move_service.ts [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+        const winner = this.checkForWinner(board)
+        if (winner !== null) return scoreMap[winner]
+        if (this.checkForDraw(board)) return scoreMap.tie
+
+        for (let move = 0; move < board.length; move += 1) {
+            if (board[move] === null) {
```
Comment (andremleblanc) [inline packages/server/src/process_move_service.ts:25]: Oh, one thing I noticed was that the styles don't like us using `continue` but that can result in an `if` like this where the entire contents of the `for` loop are nested.

Alternatively, I could have added an early bail like `if (board[move] !== null) continue` and then the rest of the for loop wouldn't be nested.

File: packages/server/src/process_move_service.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+    call() {
+        const { board } = this
+        let winner = ProcessMoveService.checkForWinner(board)
+
+        if (winner) return { board, winner }
+        if (ProcessMoveService.checkForDraw(board)) return { board, winner: Winner.tie}
+
+        const move = this.determineBestMove()
+        board[move as number] = Marker.o
```
Comment (andremleblanc) [inline packages/server/src/process_move_service.ts:92]: This is another spot where I was trying to get the compiler to trust me that move would be a number if this code is executed.

File: packages/shared/src/constants.ts [hunk 11]
```diff
@@ -6,8 +6,8 @@ export enum Difficulty {
 }
 
 export enum Marker {
-  o = 'o',
-  x = 'x',
+  o = 'O',
```
Comment (andremleblanc) [inline packages/shared/src/constants.ts:9]: Purely so the letters would be capitalized, alternatively could have transformed through CSS.

File: packages/shared/src/types.ts [hunk 12]
```diff
@@ -13,7 +13,7 @@ export interface WinningLine {
 
 export type GameStatus = {
   board: Board;
-  winner: keyof typeof Winner | null;
+  winner: typeof Winner.o | typeof Winner.x | typeof Winner.tie | null;
```
Comment (andremleblanc) [inline packages/shared/src/types.ts:16]: Consequence of changing the value in the enum, but felt less coupled. Not super confident in this change vs the CSS transform.
